### DISCUSSION TITLE
Weird speed difference for a simple increment operation of public struct field

### DISCUSSION THREAD

**Discussion by JHBonarius** [dotnet/roslyn#59030] [Q&A]
Somebody posted [this](https://twitter.com/badamczewski01/status/1484990815910957058) on Twitter. It's a simple increment loop. Let's say the code looks like this:
```csharp
using System;
public class C {
    struct MyStruct{
        public int X;
    }
    
    public int Slow() {
        var mstr = new MyStruct();
        for (int i=0; i<1000000; ++i){
            mstr.X += 1;
        }
        return mstr.X;
    }
    public int Fast() {
        var mstr = new MyStruct();
        for (int i=0; i<1000000; ++i){
            mstr.X = mstr.X + 1;
        }
        return mstr.X;
    }
}
```
He did a performance test on this, and although the two loops effectively do the same, the "Fast" loop is almost **twice** as fast as the slow loop.
![afbeelding](https://user-images.githubusercontent.com/36847155/150670987-f182236f-da9b-4887-b9b9-e6ebd7908fef.png)

I've put this code in [Sharplab.IO](https://sharplab.io) and got the following IL differences:
![afbeelding](https://user-images.githubusercontent.com/36847155/150671242-7b328ae0-bbd8-457b-8a84-89c3f01d18fd.png)

Looking at the x86 JIT assembly, I also see significant differences
![afbeelding](https://user-images.githubusercontent.com/36847155/150671297-b8ab3885-25c4-4a60-826e-f617f1d8b54e.png)

What's going on? Is this a compiler bug? The compiler should easily spot this right? It's such a common pattern.

**Comment by RikkiGibson** [comment]
It's probably a fairly complex sequence of effects resulting in a perf difference here. Comparing the JIT ASM:

```asm
C.Slow()
    L0000: push ebp
    L0001: mov ebp, esp
    L0003: sub esp, 0x18
    L0006: vxorps xmm4, xmm4, xmm4
    L000a: vmovdqu [ebp-0x18], xmm4
    L000f: xor eax, eax
    L0011: mov [ebp-8], eax
    L0014: mov [ebp-4], ecx
    L0017: cmp dword ptr [0x1c12c190], 0
    L001e: je short L0025
    L0020: call 0x6e9d4670
    L0025: nop
    L0026: xor eax, eax
    L0028: mov [ebp-8], eax
    L002b: mov [ebp-0xc], eax
    L002e: nop
    L002f: jmp short L0045
    L0031: nop
    L0032: lea eax, [ebp-8]
    L0035: mov [ebp-0x18], eax
    L0038: mov eax, [ebp-0x18]
    L003b: inc dword ptr [eax]
    L003d: nop
    L003e: mov eax, [ebp-0xc]
    L0041: inc eax
    L0042: mov [ebp-0xc], eax
    L0045: cmp dword ptr [ebp-0xc], 0xf4240
    L004c: setl al
    L004f: movzx eax, al
    L0052: mov [ebp-0x10], eax
    L0055: cmp dword ptr [ebp-0x10], 0
    L0059: jne short L0031
    L005b: mov eax, [ebp-8]
    L005e: mov [ebp-0x14], eax
    L0061: nop
    L0062: jmp short L0064
    L0064: mov eax, [ebp-0x14]
    L0067: mov esp, ebp
    L0069: pop ebp
    L006a: ret
```

```asm
C.Fast()
    L0000: push ebp
    L0001: mov ebp, esp
    L0003: sub esp, 0x14
    L0006: xor eax, eax
    L0008: mov [ebp-8], eax
    L000b: mov [ebp-0xc], eax
    L000e: mov [ebp-0x10], eax
    L0011: mov [ebp-0x14], eax
    L0014: mov [ebp-4], ecx
    L0017: cmp dword ptr [0x1c12c190], 0
    L001e: je short L0025
    L0020: call 0x6e9d4670
    L0025: nop
    L0026: xor eax, eax
    L0028: mov [ebp-8], eax
    L002b: mov [ebp-0xc], eax
    L002e: nop
    L002f: jmp short L0041
    L0031: nop
    L0032: mov eax, [ebp-8]
    L0035: inc eax
    L0036: mov [ebp-8], eax
    L0039: nop
    L003a: mov eax, [ebp-0xc]
    L003d: inc eax
    L003e: mov [ebp-0xc], eax
    L0041: cmp dword ptr [ebp-0xc], 0xf4240
    L0048: setl al
    L004b: movzx eax, al
    L004e: mov [ebp-0x10], eax
    L0051: cmp dword ptr [ebp-0x10], 0
    L0055: jne short L0031
    L0057: mov eax, [ebp-8]
    L005a: mov [ebp-0x14], eax
    L005d: nop
    L005e: jmp short L0060
    L0060: mov eax, [ebp-0x14]
    L0063: mov esp, ebp
    L0065: pop ebp
    L0066: ret
```

Seems the "slow" is deciding to use xmm instead of a mov here. I'd be curious to know what happens if we change the code to operate on an array of structs and modify the struct at `arr[i]`. cc @tannergooding in case you have insight about the observed behavior.

**Reply by tannergooding** [reply]
The `xmm` isn't the slow part, that's just the method prologue zeroing logic for stack locals. It's not even going to show up in contrast to the main loop.

The main difference is:
```asm
; slow
lea eax, [ebp-8]
mov [ebp-0x18], eax
mov eax, [ebp-0x18]
inc dword ptr [eax]
```

vs
```asm
; fast
mov eax, [ebp-8]
inc eax
mov [ebp-8], eax
```

Which is caused by `slow` emitting:
```
ldflda int32 C/MyStruct::X
dup
ldind.i4
ldc.i4.1
add
stind.i4
```

vs `fast` emitting:
```
ldloca.s 0
ldloc.0
ldfld int32 C/MyStruct::X
ldc.i4.1
add
stfld int32 C/MyStruct::X
```

The notable difference here is that in `slow` the address is being evaluated once and then duplicated (so the load is "indirect") where-as with `fast` its "copying" the value instead, which allows optimizations to light up more easily (but which may itself be "bad" in the case of larger structs if promotion doesn't happen).

Given this is a struct local (and therefore no real addressing/aliasing concerns exist) there are things both Roslyn and the JIT could do better here.

------------------------

It's also worth noting the code given in the OP is not quite the twitter logic. Note that the twitter logic uses a property, not a field.

The difference there is just the difference between `ldloca.s 0; dup` (slow) vs `ldloca.s 0; ldloca.s 0` (fast). A commit recently went in that should fix that specific scenario (https://github.com/dotnet/runtime/pull/64171) and so should also improve the field scenario.

**Reply by RikkiGibson** [reply]
Bleh, and it would probably help if I used release mode output:

```asm
C.Slow()
    L0000: push ebp
    L0001: mov ebp, esp
    L0003: push eax
    L0004: xor eax, eax
    L0006: mov [ebp-4], eax
    L0009: xor eax, eax
    L000b: lea edx, [ebp-4]
    L000e: inc dword ptr [edx]
    L0010: inc eax
    L0011: cmp eax, 0xf4240
    L0016: jl short L000b
    L0018: mov eax, [ebp-4]
    L001b: mov esp, ebp
    L001d: pop ebp
    L001e: ret

C.Fast()
    L0000: push ebp
    L0001: mov ebp, esp
    L0003: xor eax, eax
    L0005: xor edx, edx
    L0007: inc eax
    L0008: inc edx
    L0009: cmp edx, 0xf4240
    L000f: jl short L0007
    L0011: pop ebp
    L0012: ret
```
